--- at-3.1.10/at.c.shell	2006-09-12 11:25:31.000000000 +0200
+++ at-3.1.10/at.c	2006-09-12 12:25:43.000000000 +0200
@@ -460,6 +460,8 @@
     fprintf(fp, " || {\n\t echo 'Execution directory "
 	    "inaccessible' >&2\n\t exit 1\n}\n");
 
+    fprintf(fp, "${SHELL:-/bin/sh} << `(dd if=/dev/urandom count=200 bs=1 2>/dev/null|LC_ALL=C tr -d -c '[:alnum:]')`\n\n");
+
     istty = isatty(fileno(stdin));
     if (istty) {
 	fprintf(stderr, "at> ");
@@ -1037,7 +1039,7 @@
 	   It also alows a warning diagnostic to be printed.  Because of the
 	   possible variance, we always output the diagnostic. */
 
-	fprintf(stderr, "warning: commands will be executed using /bin/sh\n");
+//	fprintf(stderr, "warning: commands will be executed using /bin/sh\n");
 
 	writefile(timer, queue);
 	break;
